doc/design/proposal-moving-hooks-on-thread.md

```diff
@@ -0,0 +1,160 @@
+
+Once that is achieved, there is one large gap left to fill before we can remove the off-thread hooks: provide a way to perform async work, now that the hooks are all sync.
+
+### `startGraph` hook (module graph entry point only)
+
+We would create a new _async_ hook that runs for the main entry point, any worker entry point, and any dynamic `import()`: in other words, before any new module graph is created. This hook would delay any further hooks for that module graph operation, until the `startGraph` hook promise has resolved successfully. This provides an opportunity to do async work that the later sync hooks could have access to, for example to make network calls in a non-blocking way. One of the use cases for this hook is to support network imports, such as the [Import from HTTPS example](https://nodejs.org/api/module.html#import-from-https) in our docs. See [`nodejs/node` #43245](https://github.com/nodejs/node/pull/43245).
+
+Because this hook might need to resolve and load the entire module graph to do its work, such as to resolve and fetch HTTPS URLs, it will be provided with the `resolve` and `load` hooks in its `context`. It would also receive `context.entry` to distinguish main/worker entry points from dynamic `import()`. Its signature would be something like:
+
+```js
+/**
+ * @typedef {{
+ *   parentURL?: string,
+ *   entry: boolean,
+ *   conditions?: string[],
+ *   importAttributes?: Record<string, string>,
+ *   resolve: Function,
+ *   load: Function,
+ * }} ModuleStartGraphContext
+ */
+/**
+ * @typedef {Promise<{
+ *   shortCircuit?: boolean
+ * }>} ModuleStartGraphResult
+ */
+/**
+ * @param {string} specifier
+ * @param {ModuleStartGraphContext} context
+ * @param {(specifier: string, context: ModuleStartGraphContext) => ModuleStartGraphResult} nextStartGraph
+ * @returns {ModuleStartGraphResult}
+ */
+async function startGraph(specifier, context, nextStartGraph) {}
+```
+
+## Examples
+
+### Import from HTTPS
+
+the [Import from HTTPS example](https://nodejs.org/api/module.html#import-from-https) in our docs could use `startGraph` to handle the async work:
+
+```js
+import { parseAllImportSpecifiersFromSource } from 'some-library';
+
+/** @type {Map<URL['href'], string | Buffer} */
+const httpImportsCache = new Map();
+
+export async function startGraph(specifier, context, nextStartGraph) {
+  const { resolve, load } = context;
+  const { url } = resolve(specifier, context);
+  const urls = [url];
+  while (urls.length > 0) {
+    const url = urls.shift();
+    if (url.startsWith('https://')) {
+      const response = await fetch(url);
+      const responseSource = await response.text();
+      httpImportsCache.set(url, responseSource);
+    }
+    // load API constructs context automatically from registry data
+    const { source } = load(url);
+    httpImportsCache.set(url, source);
+    for (const specifier of parseAllImportSpecifiersFromSource(source)) {
```

Comment:
Note that only works for the top-level module and doesn't work for the graphs nested inside.

Reply:
What do you mean? This loop runs for all modules within this graph, and the hook runs again for any new graph.

Reply:
If you have

```
import('root');
```

And

```
import 'a';
import 'b';
import 'c';
```

This only runs for `root`, not for `a`/`b`/`c` since they are not in the root.

Reply:
Yes, hence the helper to parse the imports and then loop through them and resolve and load each one. It's essentially parsing the whole graph in miniature within this hook.

Reply:
It is not doing it recursively, however. But even if you implement it recursively, it can only add logic at either end of the load chain, but not in the middle.

Reply:
The user is responsible for deciding where in the chain a hook goes, by order of registration. And I’m not sure what “doing it recursively” means. This isn’t following the same evaluation order as ESM, where it goes from the leaves of the tree back to the root, but that’s not necessary for the particular use case in this example. There are other solutions such as `Atomics.wait` etc for doing async work in `load` if maintaining evaluation order is important to the particular customization use case.

Reply:
The code here does run for `'a'`, `'b'` and `'c'` because the work list gets appended with the new URLs at the end of the loop per the standard approaches for writing a recursive function as a loop.

Reply:
I think this note can be resolved? It’s doing recursion through a traditional loop, like Guy wrote.